otherlibs/systhreads/st_stubs.c

```diff
@@ -298,12 +303,11 @@ static caml_thread_t caml_thread_new_info(void)
 
 #ifndef NATIVE_CODE
   th->trap_sp_off = 1;
-  th->trap_barrier_off = 2;
+  th->trap_barrier_off = 2; /* TODO: 0? trap_barrier_block? */
   th->external_raise = NULL;
```

Comment:
@dustanddreams would you by chance be able to dive into this `trap_barrier_off` business and give an opinion on what would be best here?

For now I think that keeping the current behavior unchanged with the TODO question is okay, so we *could* merge the PR without clarification if useful to make progress, but it would of course be nicer to answer the question.

Reply:
cc also @fabbing who presented an impressive detailed account of how exception handling works, hence they might be familiar with this already.

Reply:
(The question is why is there a discrepancy—the only one—with `domain_create` here.)

Reply:
I don't know why there is such a discrepancy here. I think `caml_thread_new_info` should definitely initialize `trap_barrier_block` to `-1` here, and I also think setting `trap_barrier_off` to `0` rather than `2` in `domain_create` would make sense, unless this was done on purpose to prevent barriers from firing too early in a domain's life (in which case a comment explaining that rationale would have been nice).

Reply:
Thanks a lot for your analysis.

I see that these fields were not there in OCaml 4. They are new, including this initialization code.

So it is worth fixing quickly. A fix would need a second person to review (not me). Hence I recommend fixing it separately and leave the TODO in the code (and create a new issue about it/add it to the systhread issue).

Reply:
I agree with your analysis and am willing to merge the PR, but you need to rebase it first. This is a bugfix PR, so you should include it in the 5.2 section.

Reply:
> So it is worth fixing quickly. A fix would need a second person to review (not me). Hence I recommend fixing it separately and leave the TODO in the code (and create a new issue about it/add it to the systhread issue).

There should be two TODO then, one in `caml_thread_new_info` regarding the non-initialization of `trap_barrier_block` and one in `domain_create` regarding the initialization of `trap_barrier_off`.

Reply:
I also think `trap_barrier_block` should be initialized to `-1`.
But, if my understanding is correct, `trap_barrier_block` should stay initialized to `2`, to avoid triggering a trap when reaching the default exception handler (as guessed from https://github.com/gadmm/ocaml/blob/f00e977cc5779fa73c42a0db5bcb112de7f16237/runtime/debugger.c#L248). Also this constant is expected in this computation: https://github.com/gadmm/ocaml/blob/f00e977cc5779fa73c42a0db5bcb112de7f16237/runtime/interp.c#L157

Reply:
@dustanddreams As far I can see, `trap_barrier_off` is already initialized to 0 inside `domain_create`. Did you mean something different?

Reply:
Thanks @fabbing and @dustanddreams, I'll simply keep the TODO as is for now.

Reply:
This issue is now recorded as part of #12399 with a link to the present discussion